rfcs/20180817-variables-20.md

```diff
@@ -0,0 +1,140 @@
+# Variables in TensorFlow 2.0
+
+| Status        | Proposed       |
+:-------------- |:---------------------------------------------------- |
+| **Author(s)** | apassos@google.com |
+| **Sponsor**   | wicke@google.com, joshl@google.com, ashankar@google.com                 |
+| **Updated**   | 2018-08-17                                           |
+
+
+## Objective
+
+The API for TensorFlow variables has many drawbacks: impossible-to-reason-about semantics, reliance on global scopes, and reliance on global collections. As the TensorFlow API moves to become more pythonic and object oriented, with the Keras layers and models and the object-based serialization, we no longer have a need for much of this global infrastructure around variables.
+
+
+## Main changes
+
+The API for Variables will then change in the following ways for TF 2.0:
+
+
+
+*   tf.Variable will become an abstract base class with a well-defined interface and a scoped factory to construct instances
+    *   users will be able to implement their own variable-like objects by subclassing tf.Variable and adding a scoped factory function to use those variables
+*   variable_scope and get_variable will be removed
+    *   the tf 1.0 version of variable_scope and get_variable will be left in tf.compat.v1
+    *   to control variable naming users can use tf.name_scope + tf.Variable
+    *   whether a variable is shared across sessions / processes will be controlled by a constructor argument to tf.Variable; no other type of scope reuse will be done in the framework
+    *   scoped partitioning will be implemented as a factory function at first
+    *   libraries and users are encouraged to reuse variables by reusing their objects, like Keras layers do
+    *   custom_getters will have the following API: [variable_creator_scope](https://github.com/tensorflow/tensorflow/blob/567189980f7a1c2aa09a5170bd8d01a6ec37d303/tensorflow/python/ops/variable_scope.py#L2402)
+*   the default implementation of the tf.Variable interface will be ResourceVariable
+    *   RefVariable will be kept in tf.compat.v1 and will be the default implementation for tf.compat.v1.Variable
+    *   tf.compat.v1.Variable will have a use_resource argument to control whether a resource variable or a ref variable will be created
+*   symbols like tf.assign* will be removed in favor of methods in tf.Variable
```

Comment:
Please make item assignment possible:
```pycon
>>> import tensorflow as tf
>>> a = tf.Variable([1, 2, 3])
>>> a[1] = 5
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'Variable' object does not support item assignment
```
* It's already possible via other methods (`tf.scatter*`) but really cumbersome. 
* Since it's already possible, I assume it can't be that hard implement (but maybe I'm missing something)
* It would make teaching TensorFlow easier ("it's just like NumPy")
* It's one of those little things that makes some people prefer PyTorch: they can say "PyTorch is just like NumPy", but it's harder to say this about TensorFlow when something as fundamental to NumPy is missing.
Please, pretty please with sugar on top?  ;-)